client/rpc-servers/src/middleware.rs

```diff
@@ -0,0 +1,80 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Middleware for RPC requests.
+
+use jsonrpc_core::{Middleware as RequestMiddleware, Metadata, Request, Response, FutureResponse, FutureOutput};
+use prometheus_endpoint::{Registry, Counter, PrometheusError, register, U64};
+
+use futures::{future::Either, Future};
+use log::error;
+
+struct RpcMetrics {
+	rpc_calls: Counter<U64>,
+}
+
+impl RpcMetrics {
+	fn register(r: &Registry, transport_prefix: &str) -> Result<Self, PrometheusError> {
+		let collector_name = format!("{}_rpc_calls_total", transport_prefix);
```

Comment:
Adding `transport_prefix` as part of the metric name implies that you only want to aggregate metrics of the same transport prefix. Say you ever want to compare metrics across transport handlers. One would prefer a label, e.g. `transport_handler` instead, as PromQL, the Prometheus query language, allows you to aggregate across labels, but not (at least not without pain) across metric names.

Does that make sense?

Reply:
(Please correct me, if I'm mistaken) So you suggest adding a common label "transport_handler" for all collectors for these middlewares?

Reply:
Yes. I would suggest keeping the metric name the same across handlers (e.g. `rpc_calls_total`) but add a label (e.g. `transport_handler`) to differentiate between handlers.

With this structure you could e.g. count all requests ever made on a machine (`instance`) via `sum(rpc_calls_total) by (instance)`.

Does that make sense?

Reply:
it is. But it seems to be not possible )) Prometheus doesn't allow registration of metrics with the same name. It was the main reason, why I introduced the prefix_name - to distinguish metrics from different handlers. It's pretty dumb actually, because I can only register, unregister or collect data from the metric. But I cannot receive the access to the already registered metric.

Reply:
> Prometheus doesn't allow registration of metrics with the same name.

Yes, that is by design. Instead of creating the `RpcMetrics` struct in `RpcMiddleware::new` I would create it in `start_rpc_servers` and pass a clone of it to each `RpcMiddleware::new` invocation.

What do you think?

Reply:
Let me try and see, how it will look (it's a little bit more complicated, because handler is created in 2 places, not in 1, so i need to create an instance of metrics "higher")

Reply:
Done